external-crates/move/crates/move-trace-format/src/memory_tracer.rs

```diff
@@ -0,0 +1,196 @@
+                    //  left on the operand stack. For the initial call, these pops may (should) fail, but that
+                    // is fine as we already have the values in the parameter list.
+                    self.operand_stack.pop();
+                    locals.insert(i, p.clone());
+                }
+                self.call_stack
+                    .insert(frame.frame_id, (locals, frame.is_native));
+            }
+            TraceEvent::CloseFrame { .. } => {
+                self.call_stack.pop_last().unwrap();
+            }
+            TraceEvent::Effect(ef) => match &**ef {
+                Effect::ExecutionError(_) => (),
+                Effect::Push(value) => {
+                    self.operand_stack.push(value.clone());
+                }
+                Effect::Pop(_) => {
+                    self.operand_stack.pop().unwrap();
+                }
+                Effect::Read(Read {
+                    location,
+                    root_value_read: _,
+                    moved,
+                }) => {
+                    if *moved {
+                        match location {
+                            Location::Local(frame_idx, idx) => {
+                                let frame = self.call_stack.get_mut(frame_idx).unwrap();
+                                frame.0.remove(idx);
+                            }
+                            Location::Indexed(..) => {
+                                panic!("Cannot move from indexed location");
+                            }
+                            Location::Global(..) => {
+                                panic!("Cannot move from global location");
+                            }
+                        }
+                    }
+                }
+                Effect::Write(Write {
+                    location,
+                    root_value_after_write: value_written,
+                }) => match location {
+                    Location::Local(frame_idx, idx) => {
+                        let frame = self.call_stack.get_mut(frame_idx).unwrap();
+                        frame.0.insert(*idx, value_written.clone());
+                    }
+                    Location::Indexed(location, _idx) => {
+                        let val = self.get_mut_location(location);
+                        *val = value_written.clone().snapshot().clone();
+                    }
+                    Location::Global(id) => {
+                        let val = self.loaded_state.get_mut(id).unwrap();
+                        *val = value_written.snapshot().clone();
+                    }
+                },
+                Effect::DataLoad(DataLoad {
+                    location, snapshot, ..
+                }) => {
+                    let Location::Global(id) = location else {
+                        unreachable!("Dataload by reference must have a global location");
+                    };
```

Comment:
This is a universal property -- although it _may_ be a reference pointing higher up (we just don't know/have no way of recovering this without very significant effort). Because of this we always view references coming from natives as being "rooted" at the global index (and post processing can try and disambiguate/make those locations transparent).

I might keep the location as-is (as a `Location`) so that in the future we _can_ try to make references returned from native functions transparent, but at least for now they're opaque (and therefore this is always a Global index).

Reply:
And that can happen if a native function was passed in a reference and then derived a returned reference from it, right? Any case, makes sense, thanks for explaining!